File: java/src/LcOffer041.java
import java.util.LinkedList;
import java.util.Queue;

public class LcOffer041 {
	public static void main(String[] args) {
		MovingAverage2 movingAverage2 = new MovingAverage2(3);
		System.out.println(movingAverage2.next(1));
		System.out.println(movingAverage2.next(10));
		System.out.println(movingAverage2.next(3));
		System.out.println(movingAverage2.next(5));
	}
}

class MovingAverage2 {
	Queue queue;
	int size;
	int capacity;
	double sum;
	/** Initialize your data structure here. */
	public MovingAverage2(int size) {
		queue = new LinkedList<>();
		this.capacity = size;
		this.size = 0;
		sum = 0;
	}

	public double next(int val) {
		if (this.size < capacity) {
			queue.add(val);
			this.size++;
		} else {
			int t = (int)queue.poll();
			queue.add(val);
			sum -= t;
		}
		sum += val;
		return sum / this.size;
	}

}

File: GoLang/solution/lcOffer041.go
package solution

type MovingAverage struct {
	num []int
	size int
	sum float64
}


/** Initialize your data structure here. */
func ConstructorOffer041(size int) MovingAverage {
	return MovingAverage{make([]int, 0), size, 0}
}


func (this *MovingAverage) Next(val int) float64 {
	length := len(this.num)

	if length >= this.size {
		t := this.num[0]
		this.sum -= float64(t)
		this.num = this.num[1:]
		length = this.size
	} else {
		length++
	}
	this.num = append(this.num, val)
	this.sum += float64(val)
	return this.sum / float64(length)
}

